File: files/CH-207.R
library(tidyverse)
library(readxl)

path = "files/CH-207 Custom Grouping.xlsx"
input = read_excel(path, range = "B2:C18")
test  = read_excel(path, range = "B2:D18")

initial_threshold <- 100
threshold_step <- 50

result = input %>%
  mutate(
    Group = accumulate(
      Sales,
      .init = list(total = 0, Group = 1),
      .f = function(acc, x) {
        threshold <- initial_threshold + (acc$Group - 1) * threshold_step
        if (acc$total + x > threshold) {
          list(total = x, Group = acc$Group + 1)
        } else {
          list(total = acc$total + x, Group = acc$Group)
        }
      }
    )[-1] %>%
      map_int("Group")
  )

all.equal(result, test)
#> [1] TRUE

File: files/200-299/207/CH-207.py
import pandas as pd

path = "CH-207 Custom Grouping.xlsx"
input = pd.read_excel(path, usecols="B:C", skiprows=1, nrows=17)
test = pd.read_excel(path, usecols="B:D", skiprows=1, nrows=17)

def accumulate_groups(sales, threshold=100, step=50):
    total, group, groups = 0, 1, []
    for x in sales:
        if total + x > threshold:
            group, total = group + 1, x
            threshold += step
        else:
            total += x
        groups.append(group)
    return groups

input["Group"] = accumulate_groups(input["Sales"])

print(input["Group"].equals(test["Group"])) # True
